qt/NativeIcon.cc

```diff
@@ -0,0 +1,148 @@
+// This file Copyright © Mnemosyne LLC.
+// It may be used under GPLv2 (SPDX: GPL-2.0-only), GPLv3 (SPDX: GPL-3.0-only),
+// or any future license endorsed by Mnemosyne LLC.
+// License text can be found in the licenses/ folder.
+
+#include "NativeIcon.h"
+
+#include <iostream>
+#include <set>
+
+#include <QFontInfo>
+#include <QOperatingSystemVersion>
+#include <QPainterPath>
+#include <QRawFont>
+#include <QScreen>
+#include <QTextLayout>
+#include <QtGui/QIcon>
+#include <QtGui/QPixmap>
+#include <QtGui/QPainter>
+#include <QtGui/QFont>
+#include <QtGui/QGuiApplication>
+#include <QtGui/QPalette>
+
+namespace
+{
+
+#if defined(Q_OS_WIN)
+QPixmap makeFluentPixmap(QChar const codepoint, int const point_size)
+{
+    static auto const is_win11 = QOperatingSystemVersion::current() >= QOperatingSystemVersion::Windows11;
+    auto const fontName = is_win11 ? QStringLiteral("Segoe Fluent Icons") : QStringLiteral("Segoe MDL2 Assets");
+    auto const font = QFont{fontName, point_size};
+
+    if (!QFontMetrics{font}.inFont(codepoint))
+	    return {};
+
+    // TODO: HDPI, pixel size vs point size?
+    auto const rect = QRect{ QPoint{}, QSize{point_size, point_size}};
+    auto pixmap = QPixmap{ rect.size() };
+    pixmap.fill(Qt::red);
+    auto painter = QPainter{ &pixmap };
+    painter.setFont(font);
+    painter.setBrush(Qt::NoBrush);
+    painter.setPen(QPen{Qt::green});
+    auto br = QRect{};
+    painter.drawText(rect, Qt::AlignCenter, QString{codepoint}, &br);
+    painter.end();
+
+    std::cerr << "br x " << br.x() << " y " << br.y() << " w " << br.width() << " h " << br.height() << std::endl;
+    // FIXME: segoe icon not rendering -- br.width() is zero??
+    // This is a blocker
```

Comment:
I might need a lifeline on this one. If someone understands what the issue here is I would *greatly* appreciate the assist

Reply:
related: 

1. https://github.com/AlienCowEatCake/ImageViewer/blob/bd3203965fa77a0d50b742d870d57f844f3eff84/src/QtUtils/src/Utils/ThemeUtils.cpp#L656 from https://github.com/AlienCowEatCake/ImageViewer/commit/5885065f19403cf7a37654ff78288e8cce32d076

2. https://github.com/3360702506/QWinUI/blob/bb512b5aac9a715297d1336e175e94303fb9b221/QWinUI/src/Controls/QWinUIDropDownButton.cpp#L58 which also handles dark/light mode